### ISSUE TITLE
Nested SubResources with compound identities?

### ISSUE THREAD

**Issue by mkrinke** [api-platform/core#2025]
Hello,

first of all i would like to thank you guys for this great project. Im now into api-platform for about a week, and i very much like the project.
Basic functionalities work very well and are also documentated, but now im beginning to try implementing more complicated stuff which gives me some headaches. Hopefully you can enlighten me whether im using APIP wrong, or the features just dont exist, or... ;)

Given the following Object Structure:
* ObjectA is a Doctrine Entity and has a collection of ObjectB
* ObjectB is a POPO and has a collection ob ObjectC
* ObjectC is a POPO and has a collection of ObjectD
* ObjectD is a POPO

(ObjectA's ObjectB Tree is hydrated using a doctrine onLoad event)

I would like to implement the following routes: (GET only)
1. `/api/objectas` - List all ObjectAs, but not in great detail
2. `/api/objectas/{id}` - Get single ObjectA in great detail, include ObjectsBs in great detail, include ObjectCs with limited details
3. `/api/objectas/{id}/objectbs/{objectbs}/objectcs/{objectcs}` - Get single ObjectC in great detail, include ObjectDs with limited details
4. `/api/objectas/{id}/objectbs/{objectbs}/objectcs/{objectcs}/objectds/{objectds}` - Get single ObjectD in full detail

I used SubResource() Annotation to model the relations between the objects starting from ObjectA.

So far i stumpled upon the following problems
1. itemOperations are not generated for subResources, instead only "top level" itemOperations exist. (e.g. `/api/ObjectD/{id}`). But since ObjectB to D's identity is not global and depends on their parent, loading objects this way will not work
2. overriding itemOperations with custom path works, but the DataProvider dont receive the extra parmeters, since getItem does only support a single?! parameter
3. using RequestStack in getItem only helps so far to fetch and return the object, but when the serializer tries to generate a route it fails, since it only knows about a single parameter
4. Using compound identifiers (e.g. having methods on objectD returing the identify ob object C to A and named after the respective property) works only half way, the compound identity is detected and passed from IriConverter to the UrlGenerator, but APIP seems to have not registered the extra route parameters/slug with the route, so the generator still breaks with an exception

SubresourceDataProviderInterface's getSubresource is only called for collections, but works great for them. Thing is i dont really need collections, since for routes 2-4 im always returing a specific object containing a collection of children with limited detail (for lists/overviews, when someones clicks on an item ill load extra details)

My question now is if this concept is supported and i just did not get how to implement it from reading the documentation and code, or if there is a better way to achieve similiar results, or any other feedback you are able to provide.

Thanks!

**Comment by temp** [comment]
Same problem here. I have a simple collection + item, /customer and /customer/123, works like a charm.
Now I want to expose unloading points, with path /customer/123/unloading-points.
Unloading point has a compount key, $customerId and $code.
Since an item always needs a get resource, I tried to a lot to add a route like /customer/{customerId}/unloading-points/{code}, but can't get it working. The closest thing I got to is adding this route manually, but this fails, since the IriConverter tries to create a route with {id: customerId=123;code=001}. And this fails, since the route doesn't need the parameter id, but customerId and code. I tried a little hack in IriConverter::generateIdentifiersUrl, so that the parameters look like this: {customerId=123, code=001, id:customerId=123;code=001} and it works fine. But I don't know which side effects this would result in.

Can someone give me a hint how to solve this?
I don't want a route like /unloading-points/{id}.

Btw, I'm not using Doctrine ORM, only DBAL with models (which works really fine, except for the subresource-stuff).

**Comment by soyuka** [comment]
Could you try to decorate IriConverter or the Normalizer directly?

**Comment by temp** [comment]
Guess this would work, but isn't this something that should work out-of-the-box?

**Comment by soyuka** [comment]
Not if `UnloadingPoint` isn't a resource. If you want a subresource, by definition the subresource should be a resource (ie: have a `/unloading-points/{id}` route). 

I agree though that I often don't need this route but still want to access it's IRI. There might be something we can do to fix this but it isn't really an issue, it's a missing feature to me.

**Comment by IDSPlus** [comment]
Same problem here.

we have a resource _entityA_ with composite key and another resource _entityB_ witch is related with entityA. If we created a subresource _entityB_ from _entityA_, we get this error message "A single-valued association path expression to an entity with a composite primary key is not supported. Explicitly name the components of the composite primary key in the query".

So, a path "_/api/entityA/{id}/entityB_" with a composite _{id}_ value, it's not allowed ( by example, _id1=0;id2=0_ ).

However, if we created a custom controller for this path, simply calling the _entityB_ getter function, it works fine.

```
class EntityAController extends Controller
{

    public function entityBAction(entityA $data):PersistentCollection
    {
        return $data->getEntitydB();
    }
	
}
```


This should be totally unnecessary, so we think that this functionality should be supported directly.
